### PR TITLE
[RFC] tests: Migrate legacy test 76.

### PR DESCRIPTION
**Pull request by lucc**

### CODE REVIEW

**Comment by fwalch** [comment]
LGTM.

File: test/functional/legacy/076_completefunc_spec.lua [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    insert([=[
+      +++
+      one
+      two
+      three]=])
+
+    -- Test that nothing happens if the 'completefunc' opens.
+    -- A new window (no completion, no crash).
+    source([=[
+      function! DummyCompleteOne(findstart, base)
```
Comment (glts) [inline test/functional/legacy/076_completefunc_spec.lua]: There's something strange going on with tabs/spaces in all function bodies. The tabs aren't required here, you can replace them with spaces.

File: test/functional/legacy/076_completefunc_spec.lua [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	else
+	  return ['twodef', 'twoDEF']
+	endif
+      endfunction
+      setlocal completefunc=DummyCompleteTwo
+      /^two
+    ]=])
+    feed('A<C-X><C-U><C-N><esc>:q!<cr>')
+    -- Test that 'completefunc' works when it's OK.
+    source([=[
```
Comment (glts) [inline test/functional/legacy/076_completefunc_spec.lua]: I think using `source` is fine when it isn't overapplied.

In this test here I would extract the script-like parts (function definitions) in a `source` block at the top

```
function! DummyCompleteOne(...)
endfunction

function! DummyCompleteTwo(...)
endfunction

function! DummyCompleteThree(...)
endfunction
```

And then _below_ perform the actual test steps with normal `execute`/`feed` commands.

```
-- Test that nothing happens if the 'completefunc' opens.
-- A new window (no completion, no crash).
execute('setlocal completefunc=DummyCompleteOne')
execute('/^one')
feed('A<C-X><C-U><C-N><esc>:q!<cr>')

-- Test that ...
...
```

I hope this makes my advice about `source` a bit clearer. Perhaps somebody else can say whether they think this approach makes sense.

**Comment by glts** [comment]
> Can we use `execute` or something instead of `source`? I'd like to avoid writing a temporary file if we don't have to.

I think it was me who advised using `source`, I had seen it used in other ported tests for the 'script-like' parts of the test. I apologise if this was bad advice.

@fwalch Do you have a particular concern about the temp files (slowness, portability)?

Of course I agree that tests should be self-contained and it isn't nice that they write something to disk. Unfortunately the legacy tests are mostly _bad_ bad all around and I'm personally not sure how useful it is for us to invest our time there :|

**Comment by justinmk** [comment]
> Can we use execute or something instead of source? I'd like to avoid writing a temporary file if we don't have to.

@fwalch I've been nudging people to use source() for large vimscript blocks. Are you worried about performance? https://github.com/neovim/neovim/pull/978 has a bunch of tests on temp files vs pipes, and while the results were good for pipes (3x for some cases) this may not be the bottleneck in the tests.

It might not be a bad idea to implement an in-memory version of source()...

**Comment by fwalch** [comment]
> Are you worried about performance?

Not per se, it just seemed an unnecessary overhead. But let's just use `source` for now, it improves readability. We can change `source()` to an in-memory implementation and modify the few tests that really need something `:source`d to a new `sourcefile()` helper function or something like that at any time.

**Comment by splinterofchaos** [comment]
LGTM :+1:

**Comment by lucc** [comment]
Rebased.

Replaced leading tabs with spaces.

Left all usage of source() as is.

File: test/functional/legacy/076_completefunc_spec.lua [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  it('is working', function()
+    insert([=[
+      +++
+      one
+      two
+      three]=])
+
+    -- Test that nothing happens if the 'completefunc' opens.
+    -- A new window (no completion, no crash).
```
Comment (justinmk) [inline test/functional/legacy/076_completefunc_spec.lua]: The conversion script likes to change each line into a sentence, but it's not correct here. I'll fix before push.

**Comment by justinmk** [comment]
Merged, thanks!

**Pull request closed without merging**